ponix4-src/source/fbpanel-6.1/data/man/fbpanel.1
.\" man page originally for the Debian/GNU Linux system
.TH FBPANEL "1" "May 2005" "fbpanel 6.1" "User Commands"
.SH NAME
fbpanel \- a NETWM compliant lightweight GTK2-based panel for UNIX desktop.
.SH SYNOPSIS
.B fbpanel
[\fIOPTION\fR]
.br
.SH DESCRIPTION
.PP
fbpanel is desktop panel which provides graphical information and feedback about 
desktop activity and allows interaction with the window manager. 
It features:
.HP
\(bu taskbar \- shows a list of the managed windows (tasks)
.HP
\(bu pager \- thumbnailed view of the desktop.
.HP
\(bu launchbar \- buttons to quickly launch applications
.HP
\(bu show desktop \- button to iconify or shade all windows 
.HP
\(bu image \- display an image
.HP
\(bu clock \- show the current time and/or date
.HP
\(bu system tray \- tray for XEMBED icons (aka docklets)
.PP
fbpanel requires NETWM (www.freedesktop.org) compliant window manager. 
You can run many instances of fbpanel each with its own configuration
(see \fBOPTIONS\fR below).

Most updated info about fbpanel can be found on its home page:
http://fbpanel.sf.net/

.SH OPTIONS
.TP
\fB\--help\fR
\-- print help message and exit.
.TP
\fB\--version\fR
\-- print version and exit.
.TP
\fB\--log <number>\fR
\-- set log level 0-5. 0 - none 5 - chatty
.TP
\fB\--configure\fR
\-- open configuration dialog
.TP
\fB\--profile <name>\fR 
\-- use specified profile. The profile is loaded from the file ~/.fbpanel/<name>. 
If that fails, fbpanel will load @DATADIR@/<name>. 
No -p option is equivalent to -p default
.TP
\fB\-h\fR 
\-- same as --help
.TP
\fB\-p\fR 
\-- same as --profile
.TP
\fB\-v\fR 
\-- same as --version
.TP
\fB\-C\fR 
\-- same as --configure

.SH CUSTOMIZATION
To change default settings, copy profile file to your home directory
.br
      mkdir -p ~/.fbpanel
      cp @DATADIR@/default ~/.fbpanel
.br
and edit it. Default profile file contains comments and explanation inside,
so it should be easy. For full list of options please visit fbpanel's home page.

.SH FILES
.TP
@DATADIR@/
Directory with system-wide resources and default settings
.TP
~/.fbpanel/
Directory with the user's private profiles
.TP
~/.fbpanel/default
The user's default profile.
.SH AUTHOR
fbpanel was written by Anatoly Asviyan <aanatoly@users.sf.net>.
This manual page was originally written for the
Debian GNU/Linux system by Shyamal Prasad <shyamal@member.fsf.org>.


ponix4-const/history.txt
2011-02-03 19:45 - 2.00.575-beta - added graphical libs to usb-ip package
2011-02-04 14:45 - 2.00.576-beta -
2011-02-07 18:10 - 2.00.577-beta - build MTRR (vesa not working)
2011-02-07 18:19 - 2.00.578-beta -
2011-02-10 20:02 - 2.00.579-beta - libsane 1.0.21 with avision backend from git
2011-02-25 18:09 - 2.00.580-beta - edit libsane package
2011-02-25 18:16 - 2.00.581-beta -
2011-02-27 19:50 - 2.00.582-beta - usb-server package first build. included like package
2011-02-28 21:00 - 2.00.583-beta - final release
2011-02-28 21:07 - 2.00.584 - final release
2011-02-28 21:21 - 2.00.585 - see previos message
2011-02-28 21:45 - 2.1.001 - new version format
2011-02-28 22:51 - 2.1.1 - new version
2011-02-28 23:38 - 2.1.2 - sane package
2011-03-01 20:50 - 2.1.3 - new package sane 1.0.22
2011-03-01 23:05 - 2.1.4 -
2011-03-31 22:06 - 2.1.5 - lzma for package and usb2com converter
2011-04-18 16:44 - 2.1.6 - rxvt.pkg
2011-04-18 18:33 - 2.1.7 - rxvt.pkg ver2
2011-04-18 18:41 - 2.1.8 - rxvt ver3
2011-05-06 15:25 - 2.1.9 - repair freerdp server connect
2011-06-29 21:06 - 2.1.10 - old ver
2011-06-29 23:52 - 2.1.11 - intel video driver 2.15.0, freerdp from git, package rdesktop 1.7
2011-06-30 00:06 - 2.1.12 -
2011-06-30 00:34 - 2.1.13 -
2011-06-30 13:27 - 2.1.14 -
2011-06-30 14:20 - 2.1.15 -
2011-06-30 15:14 - 2.1.16 -
2011-06-30 15:40 - 2.1.17 -
2011-06-30 16:04 - 2.1.18 -
2011-07-01 00:11 - 2.1.19 -
2011-07-01 01:15 - 2.1.20 -
2011-07-05 20:51 - 2.1.21-standalone - rename to standalone version
2011-07-06 14:16 - 2.1.22 - add intel driver 2.15.0 and last version libdrm
2011-07-06 18:37 - 2.1.23 - rdesktop+keymaps
2011-07-07 16:05 - 2.1.24 - add keymaps to the freerdp17 package
2011-07-07 19:16 - 2.1.25 - error in the rdesktop package
2011-07-07 19:50 - 2.1.26 - too
2011-07-07 20:10 - 2.1.27 - vmview auto start
2011-07-07 20:16 - 2.1.28 - too
2011-07-07 21:08 - 2.1.29 - edit vmopenclient
2011-07-08 00:11 - 2.1.30 - edit vmopenclient
2011-07-12 21:18 - 2.1.31 - etoken
2011-07-13 18:28 - 2.1.32 - repair sane package
2011-07-13 19:02 - 2.1.33 - add smart-card support to rdesktop17 package
2011-07-21 13:38 - 2.1.34 - last version of realtek ethernet driver r8101
2011-07-21 14:52 - 2.1.35 - add firmware fo rtl8105e
2011-07-21 19:18 - 2.1.36 - new version of realtek r8169 driver (15)
2011-07-21 20:55 - 2.1.37 - without r8101.ko
2011-07-21 21:27 - 2.1.38 - r8169 from kernek, r8101 delete
2011-07-22 16:45 - 2.1.39 - add citrix client package
2011-07-22 16:51 - 2.1.40 -
2011-07-22 19:03 - 2.1.41 -
2011-07-22 19:59 - 2.1.42 -
2011-07-22 20:07 - 2.1.43 -
2011-07-22 20:40 - 2.1.44 -
2011-07-23 14:11 - 2.1.45-standalone - trying wifi build with odinary /etc/init.d/network (no ethtool)
2011-07-26 13:16 - 2.1.46 -
2011-07-26 15:20 - 2.1.47 -
2011-07-26 16:22 - 2.1.48 -
2011-07-27 14:30 - 2.1.49 -
2011-07-27 15:43 - 2.1.50 -
2011-07-27 18:05 - 2.1.51 -
2011-08-01 13:59 - 2.1.52 -
2011-08-01 15:36 - 2.1.53 -
2011-08-01 16:25 - 2.1.54 -
2011-08-01 18:18 - 2.1.55 -
2011-08-01 20:09 - 2.1.56 -
2011-08-01 20:10 - 2.1.57 - renew xorg server and libXfont
2011-08-01 20:43 - 2.1.58 -
2011-08-01 21:40 - 2.1.59 -
2011-08-01 22:32 - 2.1.60 -
2011-08-01 23:24 - 2.1.61 -
2011-08-01 23:37 - 2.1.62 -
2011-08-02 13:53 - 2.1.63 -
2011-08-02 15:07 - 2.1.64 -
2011-08-02 19:31 - 2.1.65 -
2011-08-02 20:30 - 2.1.66 -
2011-08-02 21:16 - 2.1.67 -
2011-08-02 22:29 - 2.1.68 -
2011-08-02 22:37 - 2.1.69 -
2011-08-03 16:08 - 2.1.70 -
2011-08-03 19:12 - 2.1.71 -
2011-08-03 19:47 - 2.1.72 - ICA 11 packages working!
2011-08-03 20:34 - 2.1.73 -
2011-08-03 20:43 - 2.1.74 -
2011-08-10 22:32 - 2.1.75 - add no_tls and no_nla to freerdp parameters
2011-08-11 00:30 - 2.1.76-standalone -
2011-08-11 18:35 - 2.1.77 - repair bug with freerdp_no_nla freerdp_no_tls
2011-08-11 21:45 - 2.1.78 - repair bug with Xorg (absent libfreetype)
2011-08-12 15:15 - 2.1.79 - add relatime parameter to udev script
2011-08-16 15:07 - 2.1.80 -
2011-08-16 17:26 - 2.1.81 -
2011-08-16 17:51 - 2.1.82 -
2011-08-16 21:43 - 2.1.83 - add gstreamer for ica11 package
2011-08-16 22:31 - 2.1.84 - add gstreamer and plugins for ica11 package
2011-08-18 13:57 - 2.1.85 - add gstream for speedscreen function for citrix client
2011-08-18 14:47 - 2.1.86 - add oss support to the alsa drivers
2011-08-18 15:11 - 2.1.87 -
2011-08-18 16:08 - 2.1.88 - rebuild libasound
2011-08-18 18:01 - 2.1.89 - error, soundcore must be from the kernel
2011-08-18 18:55 - 2.1.90 -
2011-09-12 18:06 - 2.1.91 - correct error with rdesktop17 package add pcsclite library
2011-09-12 22:31 - 2.1.92 - add icewm
2011-09-12 23:00 - 2.1.93 - add graphical libs to icewm
2011-09-12 23:11 - 2.1.94 - add locales to icewm
2011-09-12 23:37 - 2.1.95 -
2011-09-12 23:45 - 2.1.96 -
2011-09-13 00:07 - 2.1.97 -
2011-09-13 00:22 - 2.1.98 -
2011-09-13 13:32 - 2.1.99 -
2011-09-13 14:17 - 2.1.100 - 1
2011-09-13 19:30 - 2.1.101 -
2011-09-13 20:01 - 2.1.102 -
2011-09-13 20:40 - 2.1.103 -
2011-09-13 23:41 - 2.1.104 -
2011-09-14 15:00 - 2.1.105 -
2011-09-14 15:37 - 2.1.106 -
2011-09-16 15:08 - 2.1.107 -
2011-09-26 21:38 - 2.1.108 - in network_script added SERVER_IP=$serverid. hack to use dhcp server as tftp when no siaddr specified (current udhcpc has a bug and doesn't set tftp option)
2011-09-27 20:49 - 2.1.109 - last version of nvidia xorg driver
2011-09-28 14:48 - 2.1.110 - add --no-nla parameter for freerdp client
2011-10-03 20:00 - 2.1.111 - new version of atheros ethernet driver atl1e (AR81Family-linux-1.0.1.14) instead atl1c and atl1e (now removed)
2011-10-03 20:24 - 2.1.112-standalone - current update
2011-10-03 20:52 - 2.1.113 -
2011-10-03 21:22 - 2.1.114 -
2011-10-03 22:10 - 2.1.115 - udev-debug
2011-10-03 22:56 - 2.1.116 - udev debug
2011-10-03 23:23 - 2.1.117 - normal udev
2011-10-03 23:36 - 2.1.118 - udev debug
2011-10-03 23:42 - 2.1.119 - udev debug
2011-10-04 14:48 - 2.1.120 - corrected bug with network_script
2011-10-04 18:28 - 2.1.121 - add atl1c
2011-10-04 18:41 - 2.1.122 -
2011-10-04 19:32 - 2.1.123 - rollback to 1.0.0.9 atl1e driver

ponix4-src/source/fbpanel/fbpanel-1.0/fbpanel.c
/* fbpanel.c -- main layout engine of fbpanel project
 *
 * Copyright (C) 2002 Anatoly Asviyan <aanatoly@users.sf.net>
 *
 * This file is part of fbpanel.
 *
 * fbpanel is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * fbpanel is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with sawfish; see the file COPYING.   If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <stdlib.h>
#include <locale.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/xpm.h>


#include "fbpanel.h"

char version[] = "1.0";

#define CLOCK_CONFIG     "clock"
#define LAUNCHBAR_CONFIG "launchbar"
#define TASKBAR_CONFIG   "taskbar"
#define CLIP_CONFIG      "clip"

//#define DEBUG
#ifdef DEBUG
#define DBG(format, args...)  fprintf (stderr, format , ## args)
#else
#define DBG(args...)
#endif



char *atom_names[] = {
    "KWM_WIN_ICON",
    "_MOTIF_WM_HINTS",
    "_WIN_HINTS",
    "_WIN_LAYER",
    "_WIN_STATE",
    "WM_STATE",
    "WM_CLASS",
    
    "_NET_CLIENT_LIST",
    "_NET_NUMBER_OF_DESKTOPS",
    "_NET_CURRENT_DESKTOP",
    "_NET_DESKTOP_NAMES",
    "_NET_ACTIVE_WINDOW",

    "_NET_WM_STATE",
    "_NET_WM_STATE_SKIP_TASKBAR",
    "_NET_WM_STATE_SKIP_PAGER",
    "_NET_WM_STATE_STICKY", 
    "_NET_WM_STATE_HIDDEN",
    
    "_NET_WM_DESKTOP",
    "_NET_WM_NAME",

    "_NET_WM_ALLOWED_ACTIONS",
    "_NET_WM_ACTION_CLOSE",
    "_WIN_WORKSPACE_NAMES",
    "_NET_WM_STRUT",
};

#define ATOM_COUNT (sizeof (atom_names) / sizeof (atom_names[0]))

Atom atoms[ATOM_COUNT];

GtkWidget *window, *hbox;

Window topxwin;



//#define MAX(x,y) (((x) > (y)) ? (x) : (y))

gint delete_event(GtkWidget *, GdkEvent *, gpointer);
gint destroy_event(GtkWidget *, GdkEvent *, gpointer);



gchar *
myfgets(gchar *str, int num, FILE *fp)
{
    while (fgets(str, num, fp)) {
	g_strstrip(str);
	if (str[0] == '#' || str[0] == 0)
	    continue;
	else
	    return str;
    }
    return NULL;
}


void *
get_xaproperty (Window win, Atom prop, Atom type, int *nitems)
{
    Atom type_ret;
    int format_ret;
    unsigned long items_ret;
    unsigned long after_ret;
    unsigned char *prop_data;

    prop_data = NULL;

    XGetWindowProperty (GDK_DISPLAY(), win, prop, 0, 0x7fffffff, False,
                        type, &type_ret, &format_ret, &items_ret,
                        &after_ret, &prop_data);    
    if (nitems)
        *nitems = (prop_data) ? items_ret : 0;
    /* must be freed with XFree(prop_data) */
    return prop_data;
}


char **
get_textproperty(Window win, Atom prop, int *nitems)
{
    XTextProperty text_prop;
    char **list;
    int ret;
    
    if (XGetTextProperty(GDK_DISPLAY(), win, &text_prop, prop)) {
	ret = XmbTextPropertyToTextList(GDK_DISPLAY(), &text_prop, &list, nitems);
	XFree(text_prop.value);
	if ((ret == Success) && (*nitems > 0) && *list) {
	    /* must be freed with XFreeStringList(list); */
	    return list;
	} 
    }
    return NULL;
}


void panel_props()
{
    Atom state[3];
    unsigned int val;
    GtkRequisition requisition;
    int ret, pos;
    XWMHints wmhints;    
    unsigned int strut_new[4];
       
    gtk_widget_size_request (window, &requisition);
    gdk_window_set_decorations(window->window, 0);
    gtk_widget_set_usize(window, gdk_screen_width(), requisition.height);
    gtk_window_set_policy(GTK_WINDOW(window), FALSE, FALSE, FALSE);
	
    memset(strut_new, 0, sizeof(strut_new));
    pos = 3; /* bottom */
    strut_new[pos] = requisition.height;
    XChangeProperty (GDK_DISPLAY(), topxwin, atom__NET_WM_STRUT, XA_CARDINAL,
                     32, PropModeReplace, (unsigned char *) strut_new, 4);
    
    state[0] = atom__NET_WM_STATE_STICKY;
    state[1] = atom__NET_WM_STATE_SKIP_PAGER;
    state[2] = atom__NET_WM_STATE_SKIP_TASKBAR;
    ret = XChangeProperty (GDK_DISPLAY(), topxwin, atom__NET_WM_STATE, XA_ATOM, 32,
                           PropModeReplace, (unsigned char *) state, 3);
    
    if (ret != Success)
        DBG("XChangeProperty(%d)  failed. Error is %d\n",__LINE__, ret);

  
    val = WIN_STATE_STICKY | WIN_STATE_FIXED_POSITION;
    ret = XChangeProperty (GDK_DISPLAY(), topxwin, atom__WIN_STATE, XA_CARDINAL, 32,
                           PropModeReplace, (unsigned char *) &val, 1);
    if (ret != Success)
        DBG( "XChangeProperty(%d)  failed. Error is %d\n",__LINE__, ret);

    val = WIN_HINTS_SKIP_FOCUS | WIN_HINTS_SKIP_WINLIST | WIN_HINTS_SKIP_TASKBAR
	| WIN_HINTS_DO_NOT_COVER;
    ret = XChangeProperty (GDK_DISPLAY(), topxwin, atom__WIN_HINTS, XA_CARDINAL, 32,
                           PropModeReplace, (unsigned char *) &val, 1);
    if (ret != Success)
        DBG( "XChangeProperty(%d) failed. Error is %d\n", __LINE__, ret);

    /* make our window unfocusable */
    wmhints.flags = InputHint;
    wmhints.input = 0;
    /*XSetWMHints (GDK_DISPLAY(), topxwin, &wmhints);*/
    XChangeProperty (GDK_DISPLAY(), topxwin, XA_WM_HINTS, XA_WM_HINTS,
                     32, PropModeReplace,
                     (unsigned char *) &wmhints, sizeof (XWMHints) / 4);


}

void panel_props2()
{

    XClientMessageEvent xev;
    Atom state[3];
 
    
    xev.type = ClientMessage;
    xev.window = topxwin;
    xev.message_type = atom__NET_WM_DESKTOP;
    xev.format = 32;
    xev.data.l[0] = 0xFFFFFFFF;
    XSendEvent (GDK_DISPLAY(), GDK_ROOT_WINDOW(), False,
                (SubstructureNotifyMask|SubstructureRedirectMask), (XEvent *) &xev);

    state[0] = atom__NET_WM_ACTION_CLOSE;
    XChangeProperty (GDK_DISPLAY(), topxwin, atom__NET_WM_ALLOWED_ACTIONS,
                     XA_ATOM, 32, PropModeReplace, (unsigned char *) state, 1);
   
 }
    
void
panel_pos()
{
    GtkRequisition requisition;

    gtk_widget_size_request (window, &requisition);
    gdk_window_move_resize (window->window, 0,
                            gdk_screen_height() - requisition.height,
                            gdk_screen_width(), requisition.height);
}

void
handle_error (Display * d, XErrorEvent * ev)
{
}



static void simple_expose( GtkWidget *widget, gpointer   data )
{
    gtk_widget_queue_draw (widget);
}


void print_help()
{
    printf("fbpanel  %s\n", version);
    printf("  -v         - print version and exit\n");
    printf("  -h         - print this help and exit\n");
}

GtkWidget *separator_init()
{
    GtkWidget *sep, *eb;
    
    eb = gtk_event_box_new();
    gtk_widget_show(eb);
    gtk_signal_connect(GTK_OBJECT(eb), "expose_event",
                       GTK_SIGNAL_FUNC(gtk_widget_queue_draw), NULL);
    sep = gtk_vseparator_new();
    gtk_widget_show(sep);
    gtk_container_add (GTK_CONTAINER (eb), sep);
    gtk_widget_show(eb);
    return eb;
}

int main(int argc, char *argv[])
{
    GtkWidget *ret;
    int i;
    gchar *f;
    FILE *fp;
    
    for (i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "-v")) {
            printf("fbpanel %s\n", version);
            exit(0);
        }
        if (!strcmp(argv[i], "-h")) {
            print_help();
            exit(0);
        }
        print_help();
        exit(1);
    }   
      
    setlocale(LC_CTYPE, "");
    gtk_set_locale ();
    gtk_init(&argc, &argv);
    XSetLocaleModifiers("");
    XInternAtoms (GDK_DISPLAY(), atom_names, ATOM_COUNT, False, atoms);
    XSetErrorHandler ((XErrorHandler) handle_error);
    
    window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_container_set_border_width(GTK_CONTAINER(window), 0);
    gtk_signal_connect(GTK_OBJECT(window), "delete_event",
	GTK_SIGNAL_FUNC(delete_event), NULL);
    gtk_signal_connect(GTK_OBJECT(window), "destroy",
	GTK_SIGNAL_FUNC(destroy_event), NULL);
    gtk_widget_realize(window);
    topxwin = GDK_WINDOW_XWINDOW (GTK_WIDGET(window)->window);
  
    
    hbox = gtk_hbox_new(FALSE, 0);
    gtk_container_add(GTK_CONTAINER(window), hbox);


    // clip 
    f = g_strdup_printf("%s/.fbpanel/%s", getenv("HOME"), CLIP_CONFIG);
    fp = fopen(f, "r");
    ret = clip_init(fp);
    if (fp)
        fclose(fp);
    g_free(f);
    if (ret) 
        gtk_box_pack_start(GTK_BOX (hbox), ret, FALSE, FALSE, 0);
    ret = separator_init();
    if (ret) 
        gtk_box_pack_start(GTK_BOX (hbox), ret, FALSE, FALSE, 0);
     
    // launchbar
    f = g_strdup_printf("%s/.fbpanel/%s", getenv("HOME"), LAUNCHBAR_CONFIG);
    fp = fopen(f, "r");
    ret = lb_init(fp);
    if (fp)
        fclose(fp);
    g_free(f);
    if (ret) 
        gtk_box_pack_start(GTK_BOX (hbox), ret, FALSE, FALSE, 0);
    ret = separator_init();
    if (ret) 
        gtk_box_pack_start(GTK_BOX (hbox), ret, FALSE, FALSE, 0);
     
    // taskbar
    f = g_strdup_printf("%s/.fbpanel/%s", getenv("HOME"), TASKBAR_CONFIG);
    fp = fopen(f, "r");
    ret = tb_init(fp);
    if (fp)
        fclose(fp);
    g_free(f);
    if (ret)
        gtk_box_pack_start(GTK_BOX (hbox), ret, TRUE, TRUE, 0);

    // clock
    f = g_strdup_printf("%s/.fbpanel/%s", getenv("HOME"), CLOCK_CONFIG);
    fp = fopen(f, "r");
    ret = clock_init(fp);
    if (fp)
        fclose(fp);
    g_free(f);
    if (ret) 
        gtk_box_pack_end(GTK_BOX (hbox), ret, FALSE, FALSE, 0);

    gtk_widget_show_all(hbox);         
    panel_props();
    gtk_widget_show_all(window);
    panel_props2();
    panel_pos();

    gtk_main();
    return 0;
}

gint delete_event(GtkWidget *widget, GdkEvent *event, gpointer data)
{
  return FALSE;
}

gint destroy_event(GtkWidget *widget, GdkEvent *event, gpointer data)
{
  gtk_main_quit();
  return FALSE;
}
